optimism/client_bedrock.go

```diff
@@ -138,3 +154,53 @@ func (tx *BedrockRPCTransaction) LoadTransaction() *bedrockTransaction {
 	}
 	return ethTx
 }
+
+func toCallArg(msg Eth.CallMsg) interface{} {
+	arg := map[string]interface{}{
+		"from": msg.From,
+		"to":   msg.To,
+	}
+	if len(msg.Data) > 0 {
+		arg["data"] = hexutil.Bytes(msg.Data)
+	}
+	if msg.Value != nil {
+		arg["value"] = (*hexutil.Big)(msg.Value)
+	}
+	if msg.Gas != 0 {
+		arg["gas"] = hexutil.Uint64(msg.Gas)
+	}
+	if msg.GasPrice != nil {
+		arg["gasPrice"] = (*hexutil.Big)(msg.GasPrice)
+	}
+	return arg
+}
+
+func (ec *Client) CallContract(ctx context.Context, call Eth.CallMsg, blockNumber *big.Int) ([]byte, error) {
+	var hex hexutil.Bytes
+	err := ec.c.CallContext(ctx, &hex, "eth_call", toCallArg(call), toBlockNumArg(blockNumber))
+	if err != nil {
+		return nil, err
+	}
+	return hex, nil
+}
+
+// Dummy implementations to ensure we can leverage bind.ContractBackend interface for reusing the TCP connection
+func (ec *Client) CodeAt(ctx context.Context, contract common.Address, blockNumber *big.Int) ([]byte, error) {
+	return nil, nil
```

Comment:
Can you instead error or panic here and all other dummy functions with 'unimplemented`? Otherwise, we won't catch issues when they're used.

Reply:
Or for an even safer API, you can create a new struct that wraps the existing implementation. But this new struct implements the dummy functions rather than the wrapped struct. This reduces the scope of accidentally calling these unimplemented functions to only where we use it for the contract backend
```go
type ContractBackendImpl struct {
  services.Client
}

func (c *ContractBackendImpl) CodeAt(context.Context, common.Address, *big.Int) ([]byte, error) {
  panic("unimplemented")
}
// and so on for the other functions
```

Reply:
good point and done